67047: sample-apiserver: Add generated protobuf defs

tristanburgess opened the pull request:
- Should resolve 'errNotMarshalable' error received when requesting
        sample-apiserver endpoints with 'Accept: application/vnd.kubernetes.protobuf'
         - Add relevant sample-apiserver packages to hack/ update and verify
        protobuf scripts

<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://git.k8s.io/community/contributors/guide#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
3. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/guide/release-notes.md
4. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What this PR does / why we need it**: This PR adds generated protobuf definitions for extension API resource types in the sample-apiserver. This serves to allow for successful request/response pairs with header "Accept: application/vnd.kubernetes.protobuf", and to serve as an example of using protobuf with extension API resource types.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #67047 

**Special notes for your reviewer**:

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
NONE
```

tristanburgess committed:
67047: sample-apiserver: Add generated protobuf defs
         - Should resolve 'errNotMarshalable' error received when requesting
        sample-apiserver endpoints with 'Accept: application/vnd.kubernetes.protobuf'
         - Add relevant sample-apiserver packages to hack/ update and verify
        protobuf scripts
         - Add generated proto files to wardle API BUILD files (for
bazel)
         - Add wardle API v1beta1 to golinter exclusion list. wardle API v1alpha1
was already on the list, so this should be correct.

CaoShuFeng commented:
/ok-to-test

fedebongio commented:
/assign-sig api-machinery

nikhita commented:
:+1: thanks @tristanburgess!

/sig api-machinery
/cc @sttts @cheftako

sttts commented on hack/verify-generated-protobuf.sh:
```diff
@@ -23,7 +23,7 @@ source "${KUBE_ROOT}/hack/lib/init.sh"
 kube::golang::setup_env
 
-APIROOTS=${APIROOTS:-pkg/api pkg/apis pkg/watch staging/src/k8s.io/apimachinery/pkg/api staging/src/k8s.io/apimachinery/pkg/apis staging/src/k8s.io/apiserver/pkg staging/src/k8s.io/api staging/src/k8s.io/metrics/pkg/apis}
+APIROOTS=${APIROOTS:-pkg/api pkg/apis pkg/watch staging/src/k8s.io/apimachinery/pkg/api staging/src/k8s.io/apimachinery/pkg/apis staging/src/k8s.io/apiserver/pkg staging/src/k8s.io/api staging/src/k8s.io/metrics/pkg/apis staging/src/k8s.io/sample-apiserver/pkg/apis}
```
Am not convinced of just adding this here. We have sample-apiserver as an example repository (https://github.com/kubernetes/sample-apiserver). It is supposed to work outside of the main k/k repo. In other words, we need proto support in https://github.com/kubernetes/code-generator/blob/master/generate-internal-groups.sh which is called from sample-apiserver.

tristanburgess replied:
Gotcha, I will look into doing that instead. Just for documentation, my line of thought was that:

*From https://github.com/kubernetes/sample-apiserver: "Where does it come from?
sample-apiserver is synced from https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/sample-apiserver. Code changes are made in that location, merged into k8s.io/kubernetes and later synced here."

* kubernetes/hack/update-generated-protobuf.sh would update/gen protobuf in kubernetes/staging/src/k8s.io/sample-apiserver which gets synced to the sample-apiserver repo.

tristanburgess replied:
Would it be sufficient for this PR to add the update/verify protobuf scripts to the sample-apiserver repo, then file a new issue to dissolve the protobuf update scripts into the code-generator repo and update/verify codegen scripts?

sttts replied:
Just do both here I would say.

tristanburgess commented:
Was not an intentional CI trigger above, just trying to rebase and sync my remote as I begin work on this.

tristanburgess committed:
67047: add protobuf support to code-generator
  - Consolidate hack/*-generated-protobuf.sh with hack/*-codegen.sh
  - Update BUILD and update make rules to no longer include
  *-generated-protobuf.sh
  - Update sample-apiserver/wardle generated protobuf
  - Differentiate between packages that support protobuf on main project
  vs subproject. Skip protobuf gen on packages that weren't in the
  original list in update-generated-protobuf.sh

tristanburgess commented:
156849f is WIP until I can figure out how verify-codegen is going to work without protoc available on CI. I figure that's why update/verify generated protobuf was doing it via docker build before, but will take some time to get that working with the other codegen components.

Also need to look into if the e2e-kops-aws test is a flake.

tristanburgess commented:
156849f is WIP until I can figure out how `verify-codegen` should work without `protoc` being available on CI. I figure that's why `update-generated-protobuf` did stuff via docker build, but will take some time to figure out how to get both protobuf gen and other codegen working together.

Also need to see if `pull-kubernetes-e2e-kops-aws` failure was due to #66542

tristanburgess committed:
67047: dockerize update-codegen
 - In go-to-protobuf, remove gogoproto tags before verifying if
 --keep-gogoproto=false (by default)
 - Move subpackage verification to update to work with dockerization
 - Align with changes made to update/verify codegen in commit ff46c0f

tristanburgess commented:
@sttts, testing performed against 3bd3b82 appeared to check out, but for the sake of getting rid of the dockerized version of the proto gen (and therefore of `update-codegen` altogether), would it be possible to add `protoc` support to the CI system?

Further, I had noticed that `code-generator/go-to-proto` had a nice --verify-only flag which performs a comparison between newly generated protobuf against the existing generated files. 

https://github.com/kubernetes/kubernetes/blob/7f23a743e8c23ac6489340bbb34fa6f1d392db9d/staging/src/k8s.io/code-generator/cmd/go-to-protobuf/protobuf/cmd.go#L87

If this flag were on all of the `code-generator` commands, it could help clean up some of these `hack/` scripts, however, I only found this flag supported on `go-to-proto`. I couldn't find any issues filed to either expand or deprecate support for this flag. Do you have any knowledge of such items or thoughts on further action involving this, in comparison to putting verification logic directly in `hack/` scripts like the below?
 https://github.com/kubernetes/kubernetes/blob/7f23a743e8c23ac6489340bbb34fa6f1d392db9d/staging/src/k8s.io/sample-apiserver/hack/verify-codegen.sh

sttts commented:
> it could help clean up some of these hack/ scripts, however, I only found this flag supported on go-to-proto

It's tricky for those cases where we create files depending on tags. I believe the generator only compare to existing files. They won't notice if a file must be removed.

tristanburgess commented:
Thank you, makes sense. My changes are review-ready, then.

sttts commented on hack/update-codegen-dockerized.sh:
```diff
@@ -0,0 +1,213 @@
+fi
+
+PROTO_PACKAGES=(
+  k8s.io/api/core/v1
```
why this static list? This should be extracted automatically.

tristanburgess replied:
This particular static list already existed prior to my commit. I had tried removing the static list to see what happens, which generated a lot of protobuf for packages that apparently don't have any generated protobuf committed to the repo. 

I'm not sure if this is valid or not, or why this is the case. I did a little digging through blames and such, but didn't turn up anything certain. To err on the safe side and preserve parity,I broke it up into two pieces from the original whitelist in root/hack/update-protobuf-dockerized.sh, detailed below. 

If you like, I can get examples of the packages that aren't part of this whitelist that would otherwise generate protobuf that is not committed to the repo.

* Entries in the original whitelist that aren't sub-packages that have their own hack scripts which call code-generator/generate-internal-groups.sh remain in this new root/hack/update-protobuf-dockerized.sh

* Entries in the original whitelist that are subpackages that have their own hack scripts which call code-generator/generate-internal-groups.sh are now part of generate-internal-groups.sh

sttts commented:
/assign @cblecker 

for the dockerization.

cblecker reviewed (commented):
Just to get the intent right, we're removing the protobuf specific script, and combining it and the codegen script together to run both steps in the same docker container?

tristanburgess commented on build/common.sh:
```diff
@@ -730,8 +730,6 @@ function kube::build::sync_to_container() {
     --filter='- /_tmp/' \
     --filter='- /_output/' \
     --filter='- /' \
-    --filter='H zz_generated.*' \
```
Correct, part of the intent is to combine the protobuf script into the codegen script. 

The reason for this is because this PR adds generated protobuf to a subpackage (sample apiserver). Sub-packages have their own hack/update-codegen which calls the code-generator package facilities. These sub-package scripts do not support protobuf generation.

Instead of adding a new protobuf generation script for each sub-package, I combined the two scripts into a single script in kubernetes/hack/, and mirror this new protobuf generation in the sub-package update-codegen scripts so that they're uniform.


The dockerization part as far as I can tell is only because protoc doesn't appear to be available on the bare CI machines, and the build docker images are configured to have protoc. I'm pretty sure if protoc could be on the bare CI machines then this script could remain un-dockerized.


Further, I had added the line above because I had to make this update-codegen script un-dockerized. What I had found was that verify-codegen.sh would fail no matter what, because the rsync config for the docker build image has the zz_generated.* files removed. Thus when we go to verify newly generated codegen against existing, the zz_generated.* files aren't on the docker image, so everything looks like it never existed before.


The above made sense to me at the time, however I have since rebuilt and re-added the line above and am not able to reproduce the issue. I'm going to continue testing for a bit and will update when I have an update.

tristanburgess replied:
Ah, I looked deeper into the issue and it makes more sense to me now.

I hadn't seen until now that some generated files are actually created via a Make rule.

Specifically, I had noticed that zz_generated.deepcopy files appeared to be updated by running verify-codegen.sh if they were out of date. With the change above, this would be reported by running verify-codegen.sh that you need to run update-codegen.sh to resolve.

This is not correct, as I've now found, since those generated files aren't handled by this system.

At least, this is my current understanding. I will reverse the above change.

tristanburgess replied:
I have amended my previous commit to reverse the above change.

I also amended to include changes from commit ff46c0f which affect hack/update-codegen.sh and hack/verify-codegen.sh and needed to be reconciled with my changes.

tristanburgess commented:
/test pull-kubernetes-verify

tristanburgess commented:
pull-kubernetes-verify appears to be related to separate issue #68510

tristanburgess commented:
/test pull-kubernetes-node-e2e

tristanburgess commented:
/test pull-kubernetes-verify
